client/go/outline/iptable/packet_listener.go

```diff
@@ -0,0 +1,319 @@
+		connAttempt, err := defaultListener.ListenPacket(forwardingContext)
+
+		if err != nil {
+			defaultConn = nil
+		} else {
+			defaultConn = connAttempt
+		}
+	}
+
+	conn := &packetConn{
+		closePacketForwardingContext: closeForwardingContext,
+		connMap:                      make(map[netip.Addr]net.PacketConn),
+		connMapLock:                  sync.Mutex{},
+		defaultConn:                  defaultConn,
+		defaultListener:              defaultListener,
+		forwardedPackets:             make(chan incomingPacket, packetQueueSize),
+		forwardingContext:            forwardingContext,
+		listenerTable:                listenerTable,
+		lastRemainingPacket:          incomingPacket{},
+		lastRemainingPacketLock:      sync.Mutex{},
+	}
+
+	if conn.defaultConn != nil {
+		conn.forwardPackets(conn.defaultConn)
+	}
+
+	return conn, nil
+}
+
+func (conn *packetConn) ReadFrom(result []byte) (n int, addr net.Addr, err error) {
+	if len(conn.lastRemainingPacket.data) > 0 {
+		conn.lastRemainingPacketLock.Lock()
+		numBytes := copy(result, conn.lastRemainingPacket.data)
+		conn.lastRemainingPacket.data = conn.lastRemainingPacket.data[numBytes:]
+		conn.lastRemainingPacketLock.Unlock()
+
+		return numBytes, conn.lastRemainingPacket.addr, nil
+	}
+
+	packet, ok := <-conn.forwardedPackets
+	if !ok {
+		return 0, nil, net.ErrClosed
+	}
+
+	numBytes := copy(result, packet.data)
+
+	conn.lastRemainingPacketLock.Lock()
+	conn.lastRemainingPacket.data = packet.data[numBytes:]
+	conn.lastRemainingPacket.addr = packet.addr
+	conn.lastRemainingPacketLock.Unlock()
+
+	return numBytes, packet.addr, nil
+}
+
+func (conn *packetConn) WriteTo(packet []byte, addr net.Addr) (numBytes int, err error) {
+	var ip netip.Addr
+	if udpAddr, ok := addr.(*net.UDPAddr); ok {
+		if parsedIP, ok := netip.AddrFromSlice(udpAddr.IP); ok {
+			ip = parsedIP
+		}
+	}
+	if !ip.IsValid() {
+		return 0, fmt.Errorf("could not parse valid IP from address %v (%T)", addr, addr)
+	}
+
+	ip = ip.Unmap()
+
+	conn.connMapLock.Lock()
+	subconn := conn.connMap[ip]
+	conn.connMapLock.Unlock()
+
+	if subconn == nil {
+		listener, ok := conn.listenerTable.Lookup(ip)
+
+		if ok {
+			subconn, err = listener.ListenPacket(conn.forwardingContext)
+
+			if err != nil {
+				return 0, err
+			}
+
+			conn.connMapLock.Lock()
+			conn.connMap[ip] = subconn
+			conn.connMapLock.Unlock()
+
+			conn.forwardPackets(subconn)
+		}
+	}
+
+	if subconn == nil && conn.defaultListener != nil {
+		subconn = conn.defaultConn
+	} else if subconn == nil {
+		return 0, fmt.Errorf("no connection found for IP %s", ip.String())
+	}
+
+	return subconn.WriteTo(packet, addr)
+}
+
+func (conn *packetConn) Close() error {
+	var errs error
+	conn.closePacketForwardingContext()
+
+	conn.connMapLock.Lock()
+	for address, subconn := range conn.connMap {
+		if err := subconn.Close(); err != nil {
+			errs = errors.Join(errs, fmt.Errorf("failed to close subconnection for %s: %w", address, err))
+		}
+		delete(conn.connMap, address)
```

Comment:
Shall we put a flag in the `packetConn` indicating the `conn` has already been closed? And return `net.ErrClosed` in `WriteTo`. Consider this race condition:

```
packetConn.Close()
        |
(connMap cleared)
        |
connMapLock.Unlock()      WriteTo() is called
        |                          |
        |               subconn added to connMap
        |                          |
       ...                        ...
```

There will be one (or more) dangling `subconn`s that won't be closed.